### PR TITLE
fix duplicate validation error for quota editing

### PR DESCRIPTION
**Pull request by pacoxu**
**What type of PR is this?**
/kind bug
/kind cleanup

Detailed reason in https://github.com/kubernetes/kubernetes/pull/98201#issuecomment-765116930

**What this PR does / why we need it**:
When adding a wrong quota like `"test": "12"` in spec.hard, there are four errors return. Two of them are duplicates.

**Which issue(s) this PR fixes**:

How to reproduce it?
Create a quota and edit it with kubectl, and add `"test": "12"` in spec.hard.
```
apiVersion: v1
kind: ResourceQuota
metadata:
  name: example
spec:
  hard:
    pods: "18"
    limits.cpu: "26"
    limits.memory: "62Gi"
    "test": "12"

```
the error message is like 
```
# resourcequotas "example" was not valid:
# * spec.hard[test]: Invalid value: "test": must be a standard resource type or fully qualified
# * spec.hard[test]: Invalid value: "test": must be a standard resource for quota
# * spec.hard[test]: Invalid value: "test": must be a standard resource type or fully qualified
# * spec.hard[test]: Invalid value: "test": must be a standard resource for quota
#
```
Errors are duplicated. 

**Special notes for your reviewer**:
I read the code of LimitRange and ResourceQuota during investigating #98071 
A typo and small loop continue problem were found during the code reading. So I fixed collectively with this issue.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by pacoxu** [comment]
? not sure which sig it belongs to.
/sig api-machinery

File: plugin/pkg/admission/limitranger/admission.go
Comment (yliaog) [inline plugin/pkg/admission/limitranger/admission.go:560]: limitType is either LimitTypeContainer or corev1.LimitTypePod, why is this 'continue' necessary?

File: pkg/registry/core/resourcequota/strategy.go
Comment (yliaog) [inline pkg/registry/core/resourcequota/strategy.go:77]: validation.ValidateResourceQuotaUpdate may return an error that is not present in the errorList, i.e. it is not duplicate, 

returning here means as long as there is any error here, validation.ValidateResourceQuotaUpdate is skipped, which is not desired.

File: plugin/pkg/admission/limitranger/admission.go
Comment (pacoxu) [reply plugin/pkg/admission/limitranger/admission.go:560]: @yliaog  thanks for your comments.

Not sure which is better.  What I want to do is like below. 

```
	for i := range limitRange.Spec.Limits {
		limit := limitRange.Spec.Limits[i]	
		limitType := limit.Type	
		// enforce container limits	
		if limitType == corev1.LimitTypeContainer {
			...
+			continue //directly to next loop;skip the next if
		} 
		if limitType == corev1.LimitTypePod {
			
		}
	}
```

or

```
	for i := range limitRange.Spec.Limits {
		limit := limitRange.Spec.Limits[i]	
		limitType := limit.Type	
		// enforce container limits	
		if limitType == corev1.LimitTypeContainer {
			...
			continue
		} 
+		else if limitType == corev1.LimitTypePod {
-		if limitType == corev1.LimitTypePod {
			
		}
	}
```

Not sure if this is needed or makes sense.

However, this is not the key part of this PR.

**Comment by pacoxu** [comment]
@yliaog  To summarize why there are duplicate messages :

1. ValidateUpdate 
	1.1 errorList := validation.ValidateResourceQuota( new object))
 		- 1.1.1 ValidateObjectMeta
 		- 1.1.2  **ValidateResourceQuotaSpec( new object ) (dup)**
 		- 1.1.3 ValidateResourceQuotaStatus( new object )
	1.2 return append(errorList, validation.ValidateResourceQuotaUpdate( new object, old object)...)
 		- 1.2.1  ValidateObjectMetaUpdate
 		- 1.2.2 **ValidateResourceQuotaSpec( new object )  (dup)**
 		- 1.2.3 Scope Validate
	

https://github.com/kubernetes/kubernetes/blob/c970a46bc1bcc100bbbfabd5c12bd4c5d87f8aea/pkg/apis/core/validation/validation.go#L5533

ValidateResourceQuotaUpdate will validate the new object's spec again.

Maybe removing this line is a better fix in my opinion.

So I updated the PR. Would you take a look again? Thanks.

File: pkg/registry/core/resourcequota/strategy.go
Comment (pacoxu) [reply pkg/registry/core/resourcequota/strategy.go]: Add comments in the comments below.

File: plugin/pkg/admission/limitranger/admission.go
Comment (pacoxu) [reply plugin/pkg/admission/limitranger/admission.go]: I think `else if` is better in this case. So I changed it.

Comment (pacoxu) [reply plugin/pkg/admission/limitranger/admission.go]: @yliaog  As I have another bugfix pr #98295 for limit range, I move the cleanup code to that PR.
So the code owner could focus on the single package.

Thanks for your comments anyway.

File: pkg/apis/core/validation/validation.go
Comment (yliaog) [inline pkg/apis/core/validation/validation.go:5533]: ObjectMeta and Spec are different, i don't think ValidateResourceQuotaSpec should be skipped here.

Comment (pacoxu) [reply pkg/apis/core/validation/validation.go:5533]: The detailed reason is in https://github.com/kubernetes/kubernetes/pull/98201#issuecomment-765116930.

The duplicate lines are Line 5533  and Line 5477. Both are spec validation.

Line 5533 here:
	allErrs = append(allErrs, ValidateResourceQuotaSpec(&newResourceQuota.Spec, field.NewPath("spec"))...)
Line 5477:
	allErrs = append(allErrs, ValidateResourceQuotaSpec(&resourceQuota.Spec, field.NewPath("spec"))...)

Comment (derekwaynecarr) [inline pkg/apis/core/validation/validation.go:5533]: ok, so the strategy calls both

ValidateResourceQuota
ValidateResourceQuotaUpdate

and both operations revalidate the spec.

I agree it seems fine to only validate the spec in ValidateResourceQuota, and keep the update operation focused on allowed updates.

**Review by derekwaynecarr — APPROVED** [review]
/approve
/lgtm

File: pkg/apis/core/validation/validation.go
Comment (yliaog) [reply pkg/apis/core/validation/validation.go:5533]: would update operation possibly update spec, hence spec needs to be validated again?

File: pkg/apis/core/validation/validation.go
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:5531]: I would actually go the other direction and drop the `resourcequotaStrategy#ValidateUpdate` strategy call to `ValidateResourceQuota`:

* `ValidateObjectMetaUpdate` is sufficient on update, we don't need to call `ValidateObjectMeta` again
* `ValidateResourceQuotaSpec` is already being called from ValidateResourceQuotaUpdate and makes sense
* `newResourceQuota.Status = oldResourceQuota.Status` is already done in `resourcequotaStrategy#PrepareForUpdate`, so the validation method shouldn't do it again (mutating an input in validation is strange anyway)
* Since updates can't change status, it doesn't make sense to validate them in `ValidateResourceQuotaUpdate` either

Comment (pacoxu) [reply pkg/apis/core/validation/validation.go]: Yes. Agree.

My first try to fix is to add check after resourcequotaStrategy#ValidateUpdate
```
// ValidateUpdate is the default update validation for an end user.
func (resourcequotaStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
	errorList := validation.ValidateResourceQuota(obj.(*api.ResourceQuota))
+	if len(errorList) != 0 {
+		return errorList
+	}
	return append(errorList, validation.ValidateResourceQuotaUpdate(obj.(*api.ResourceQuota), old.(*api.ResourceQuota))...)
}
```

Your suggestion would be more direct.

Updated the PR. Thanks for your comments @liggitt.

**Comment by pacoxu** [comment]
as this is just something that is nice-to-have
/priority backlog

**Comment by pacoxu** [comment]
/assign @liggitt
Kindly ping, as the bot comments that it only needs approval from pkg/registry/OWNERS.

**Pull request merged**
